File: Atividades A1/15 Questões/GO/Questoes15/Questao4/main.go
package main

//4. Herança Crie uma classe base Animal com métodos como som. Derive classes como Cachorro e Gato que implementam o método som de maneiras diferentes.
import "fmt"

type Animal struct {
	especie string
	nome    string
}

func (a Animal) EmitirSom() string {
	return ""
}

type Cachorro struct {
	Animal
}

func NovoCachorro(nome string) Cachorro {
	return Cachorro{
		Animal: Animal{
			especie: "cachorro",
			nome:    nome,
		},
	}
}

func (c Cachorro) EmitirSom() string {
	return "Au Au"
}

type Gato struct {
	Animal
}

func NovoGato(nome string) Gato {
	return Gato{
		Animal: Animal{
			especie: "gato",
			nome:    nome,
		},
	}
}

func (g Gato) EmitirSom() string {
	return "Miau"
}

func main() {
	cachorro := NovoCachorro("Rex")
	gato := NovoGato("Mimi")

	fmt.Printf("%s diz: %s\n", cachorro.nome, cachorro.EmitirSom())
	fmt.Printf("%s diz: %s\n", gato.nome, gato.EmitirSom())
}

File: Atividades A1/15 Questões/Python/Questoes15/Questao4/Main.py
#4. Herança Crie uma classe base Animal com métodos como som. Derive classes como Cachorro e Gato que implementam o método som de maneiras diferentes.

# Classe Animal
class Animal:
    def __init__(self, especie, nome):
        self.especie = especie
        self.nome = nome

    def emitirSom(self):
        return ''


class Cachorro(Animal):
    def __init__(self, nome):
        super().__init__('cachorro', nome)

    def emitirSom(self):
        return 'Au Au'


class Gato(Animal):
    def __init__(self,nome):
        super().__init__('gato', nome)

    def emitirSom(self):
        return 'Miau'


if __name__ == '__main__':
    cachorro = Cachorro('Rex')
    gato = Gato('Mimi')

    print(f'{cachorro.nome} diz: {cachorro.emitirSom()}')
    print(f'{gato.nome} diz: {gato.emitirSom()}')
